Further remove dependency of  Compatibility.Layout on all modern .NET MAUI controls

jfversluis opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Description of Change

Ideally we will eventually be able to remove `Compatibility.Layout` but we are still in a state where if users have overridden certain methods or are calling some properties that are still on Compatibility.Layout then the binary isn't compatible. This PR copies all of the public APIs to the subclasses so whenever users override or call those methods/properties on the base class the IL will resolved to the derived type vs the `Compatibility.Layout`

We have also implemented a property that will detect if the `Compatibility.Layout` is being used with any of our modern layouts. If it is then we just bypass all of the `Compatibility.Layout` logic so that it doesn't interfere with performance or the MAUI logic

### Issues Fixed

Fixes #28736

jfversluis commented:
@albyrock87 can't add you as a reviewer, so tagging you like this. Would be great to get your perspective here!

jfversluis commented on src/Controls/src/Core/IClippedToBoundsElement.cs:
```diff
@@ -0,0 +1,12 @@
+	/// </summary>
+	bool IsClippedToBounds { get; set; }
+
+	static void IsClippedToBoundsPropertyChanged(BindableObject bindableObject, object oldValue, object newValue);
```
```suggestion
	void IsClippedToBoundsPropertyChanged(BindableObject bindableObject, object oldValue, object newValue);
```

jsuarezruiz commented on src/Controls/src/Core/ClippedToBoundsElement.cs:
```diff
@@ -0,0 +1,8 @@
+
+static class ClippedToBoundsElement
+{
+	public static readonly BindableProperty IsClippedToBoundsProperty =
```
```
Build FAILED.

C:\a\_work\1\s\.dotnet\sdk\10.0.100-preview.4.25228.15\Microsoft.Common.CurrentVersion.targets(1890,5): warning NU1702: ProjectReference 'C:\a\_work\1\s\src\Controls\src\Core.Design\Controls.Core.Design.csproj' was resolved using '.NETFramework,Version=v4.7.2' instead of the project target framework '.NETCoreApp,Version=v10.0'. This project may not be fully compatible with your project. [C:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\.dotnet\sdk\10.0.100-preview.4.25228.15\Microsoft.Common.CurrentVersion.targets(1890,5): warning NU1702: ProjectReference 'C:\a\_work\1\s\src\Controls\src\Xaml.Design\Controls.Xaml.Design.csproj' was resolved using '.NETFramework,Version=v4.7.2' instead of the project target framework '.NETCoreApp,Version=v10.0'. This project may not be fully compatible with your project. [C:\a\_work\1\s\src\Controls\src\Xaml\Controls.Xaml.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\src\Controls\src\Core\Frame\Frame.cs(11,128): error CS0535: 'Frame' does not implement interface member 'IClippedToBoundsElement.IsClippedToBoundsPropertyChanged(BindableObject, object, object)' [C:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=netstandard2.0]
C:\a\_work\1\s\src\Controls\src\Core\RefreshView\RefreshView.cs(15,117): error CS0535: 'RefreshView' does not implement interface member 'IClippedToBoundsElement.IsClippedToBoundsPropertyChanged(BindableObject, object, object)' [C:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=netstandard2.0]
C:\a\_work\1\s\src\Controls\src\Core\Frame\Frame.cs(11,128): error CS0535: 'Frame' does not implement interface member 'IClippedToBoundsElement.IsClippedToBoundsPropertyChanged(BindableObject, object, object)' [C:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\src\Controls\src\Core\RefreshView\RefreshView.cs(15,117): error CS0535: 'RefreshView' does not implement interface member 'IClippedToBoundsElement.IsClippedToBoundsPropertyChanged(BindableObject, object, object)' [C:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net10.0]
C:\a\_work\1\s\src\Controls\src\Core\Frame\Frame.cs(11,128): error CS0535: 'Frame' does not implement interface member 'IClippedToBoundsElement.IsClippedToBoundsPropertyChanged(BindableObject, object, object)' [C:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=netstandard2.1]
C:\a\_work\1\s\src\Controls\src\Core\RefreshView\RefreshView.cs(15,117): error CS0535: 'RefreshView' does not implement interface member 'IClippedToBoundsElement.IsClippedToBoundsPropertyChanged(BindableObject, object, object)' [C:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=netstandard2.1]
    2 Warning(s)
    6 Error(s)
```

albyrock87 commented on src/Controls/src/Core/LegacyLayouts/Layout.cs:
Why are we keeping this class alive?

albyrock87 commented on src/Controls/src/Core/RefreshView/RefreshView.cs:
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.Maui.Controls
 	[ContentProperty(nameof(Content))]
 	[DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
-	public partial class RefreshView : ContentView, IElementConfiguration<RefreshView>, IRefreshView, ICommandElement
+	public partial class RefreshView : ContentView, IElementConfiguration<RefreshView>, IRefreshView, ICommandElement, IClippedToBoundsElement
```
Why don't we implement `IClippedToBoundsElement` directly in `TemplatedView` so that this class gets it for free?

albyrock87 commented on src/Controls/src/Core/ScrollView/ScrollView.cs:
```diff
@@ -31,7 +28,7 @@ public Rect LayoutAreaOverride
 				_layoutAreaOverride = value;
 				// Dont invalidate here, we can relayout immediately since this only impacts our innards
-				UpdateChildrenLayout();
+				InvalidateMeasure();
```
I see this one is already obsolete in .NET9: can we remove the property completely?

```
[Obsolete]
public Rect LayoutAreaOverride
```

PureWeen replied:
We can't remove it completely because it's part of `IScrollViewController` :-( 

So, for now we can just change this to an explicit implementation of `IScrollViewController` and at least get it removed from the `txt` files

jsuarezruiz reviewed (commented):
```
C:\a\_work\1\s\src\Controls\tests\TestCases.HostApp\Issues\Bugzilla\Bugzilla36788.cs(68,5): error CS0117: 'ContentView' does not contain a definition for 'IsClippedToBounds' [C:\a\_work\1\s\src\Controls\tests\TestCases.HostApp\Controls.TestCases.HostApp.csproj::TargetFramework=net10.0-android]
C:\a\_work\1\s\src\Compatibility\Core\src\Windows\RadioButtonRenderer.cs(59,35): error CS0117: 'RadioButton' does not contain a definition for 'PaddingProperty' [C:\a\_work\1\s\src\Compatibility\Core\src\Compatibility.csproj::TargetFramework=net10.0-windows10.0.20348.0]
C:\a\_work\1\s\src\Compatibility\Core\src\Windows\RadioButtonRenderer.cs(59,97): error CS0117: 'RadioButton' does not contain a definition for 'PaddingProperty' [C:\a\_work\1\s\src\Compatibility\Core\src\Compatibility.csproj::TargetFramework=net10.0-windows10.0.20348.0]
C:\a\_work\1\s\src\Compatibility\Core\src\Windows\RadioButtonRenderer.cs(112,43): error CS0117: 'RadioButton' does not contain a definition for 'PaddingProperty' [C:\a\_work\1\s\src\Compatibility\Core\src\Compatibility.csproj::TargetFramework=net10.0-windows10.0.20348.0]
    33 Error(s)
```

PureWeen commented on src/Controls/src/Core/ScrollView/ScrollView.cs:
```diff
@@ -171,10 +169,33 @@ public View Content
 		}
 
+		/// <inheritdoc cref="IInputTransparentContainerElement.CascadeInputTransparent"/>
+		public bool CascadeInputTransparent
```
For all these properties, even though we moved them to `InputTransparentContainerElement` we still need to add the BindableProperty to this class as well

So for all uses of this (and other areas where this pattern was followed) we still need


```C#
/// <summary>Bindable property for <see cref="CascadeInputTransparent"/>.</summary>
		public static readonly BindableProperty CascadeInputTransparentProperty = InputTransparentContainerElement.CascadeInputTransparentProperty;
```

mattleibow reviewed (dismissed):
A quick toilet review 😉

PureWeen commented on src/Controls/src/Core/ContentPresenter.cs:
```diff
@@ -1,22 +1,28 @@
-		/// <include file="../../docs/Microsoft.Maui.Controls/ContentPresenter.xml" path="//Member[@MemberName='.ctor']/Docs/*" />
+#pragma warning disable CS0067
+		[Obsolete("Use SizeChanged.")]
+		public event EventHandler LayoutChanged;
```
fixed

PureWeen commented:
/rebase

PureWeen commented:
/azp run

PureWeen commented:
/rebase

PureWeen commented:
/rebase

PureWeen commented:
/rebase

PureWeen committed:
Make Compatibility.Layout a NOOP for non compat types

# Conflicts:
#	src/Controls/src/Core/PublicAPI/net-android/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-ios/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-maccatalyst/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-windows/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/netstandard/PublicAPI.Unshipped.txt

PureWeen committed:
- fix timing on Disconnecting CV

PureWeen committed:
- fix empty view layout

PureWeen committed:
- force measure and arrange call

PureWeen committed:
- fix legacy tests

PureWeen committed:
- set frame if emptyview is just text

PureWeen committed:
Update ItemsViewController2.cs

PureWeen commented:
/rebase

PureWeen committed:
- fix test

PureWeen committed:
- make sure to invalidate measure when the content is swapped out

PureWeen commented:
/rebase

PureWeen committed:
- fix appearing from firing when the content page has been covered by a modal

PureWeen committed:
- update with latest titleview images for better test and fix refreshview to measure/layout correctly

PureWeen committed:
- adjust refreshview screen shot for new measure code

PureWeen committed:
- shipped text fixes

PureWeen committed:
- fix

PureWeen commented:
Moved PR to [here](https://github.com/dotnet/maui/pull/30672)
